src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProjectFactory.cs

```diff
@@ -92,66 +94,17 @@ public Task<VisualStudioProject> CreateAndAddToWorkspaceAsync(string projectSyst
-                _visualStudioWorkspaceImpl,
-                _dynamicFileInfoProviders,
-                _hostDiagnosticUpdateSource,
-                vsixAnalyzerProvider,
-                id,
-                displayName: projectSystemName,
-                language,
-                assemblyName: assemblyName,
-                compilationOptions: creationInfo.CompilationOptions,
-                filePath: creationInfo.FilePath,
-                parseOptions: creationInfo.ParseOptions);
-
-            var versionStamp = creationInfo.FilePath != null ? VersionStamp.Create(File.GetLastWriteTimeUtc(creationInfo.FilePath))
-                                                             : VersionStamp.Create();
-
-            await _visualStudioWorkspaceImpl.ApplyChangeToWorkspaceAsync(w =>
-            {
-                _visualStudioWorkspaceImpl.AddProjectToInternalMaps_NoLock(project, creationInfo.Hierarchy, creationInfo.ProjectGuid, projectSystemName);
-
-                var projectInfo = ProjectInfo.Create(
-                    new ProjectInfo.ProjectAttributes(
-                        id,
-                        versionStamp,
-                        name: projectSystemName,
-                        assemblyName: assemblyName,
-                        language: language,
-                        checksumAlgorithm: SourceHashAlgorithms.Default, // will be updated when command line is set
-                        compilationOutputFilePaths: default, // will be updated when command line is set
-                        filePath: creationInfo.FilePath,
-                        telemetryId: creationInfo.ProjectGuid),
-                    compilationOptions: creationInfo.CompilationOptions,
-                    parseOptions: creationInfo.ParseOptions);
-
-                // If we don't have any projects and this is our first project being added, then we'll create a new SolutionId
-                // and count this as the solution being added so that event is raised.
-                if (w.CurrentSolution.ProjectIds.Count == 0)
-                {
-                    var solutionSessionId = GetSolutionSessionId();
-
-                    w.OnSolutionAdded(
-                        SolutionInfo.Create(
-                            SolutionId.CreateNewId(solutionFilePath),
-                            VersionStamp.Create(),
-                            solutionFilePath,
-                            projects: new[] { projectInfo },
-                            analyzerReferences: w.CurrentSolution.AnalyzerReferences)
-                        .WithTelemetryId(solutionSessionId));
-                }
-                else
-                {
-                    w.OnProjectAdded(projectInfo);
-                }
-            });
+            _visualStudioWorkspaceImpl.ProjectSystemProjectFactory.SolutionPath = solutionFilePath;
+            _visualStudioWorkspaceImpl.ProjectSystemProjectFactory.SolutionTelemetryId = GetSolutionSessionId();
+
```

Comment:
can't say i love what appears to be mutation of these entities across types.  (don't have any suggestions either).

Reply:
Yeah, agreed. The holder of the ProjectSystemProjectFactory should probably move to this type actually. I'll make a note to follow up on that.